apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs

```diff
@@ -624,29 +719,33 @@ private async Task<PooledSession> CreatePooledSessionAsync(CancellationToken can
                     {
                         if (acquiredSemaphore)
                         {
-                            Parent._sessionAcquisitionSemaphore.Release();
+                            Parent._batchSessionCreateSemaphore.Release();
                         }
                     }
                 }
+                catch(Exception e)
+                {
+                    Parent._logger.Warn(() => $"Failed to batch create sessions for {_databaseName}", e);
+                    throw;
+                }
                 finally
                 {
                     // Atomically set _healthy and determine whether we were previously healthy, but only if either we've succeeded,
                     // or we failed for a reason other than cancellation. We don't want to go unhealthy just because a caller cancelled
                     // a cancellation token before we had chance to create the session.
                     if (success || !canceled)
                     {
-                        bool wasHealthy = Interlocked.Exchange(ref _healthy, success ? 1 : 0) == 1;
+                        bool wasHealthy = Interlocked.Exchange(ref _lastUnhealthyBatchSize, success ? 0 : batchSize) == 0;
                         if (wasHealthy != success)
                         {
                             Parent._logger.Info(() => $"Session pool for {_databaseName} is now {(success ? "healthy" : "unhealthy")}.");
                         }
                     }
                     // If this call failed, we can make another attempt.
-                    if (!success)
-                    {
-                        Interlocked.Decrement(ref _liveOrRequestedSessionCount);
-                    }
-                    Interlocked.Decrement(ref _inFlightSessionCreationCount);
+                    // If it succeeded update the value with the actual sessions created.
```

Comment:
This comment suggests there should be an "if" here, which there isn't.
I think probably these two comments need to be rewritten along the lines of "update the state of the world to the new reality based on how many sessions were returned, which may be 0".